File: solutions/java/0239.java
class Solution {
  public int[] maxSlidingWindow(int[] nums, int k) {
    List<Integer> ans = new ArrayList<>();
    Deque<Integer> decreasingDeque = new ArrayDeque<>();

    for (int i = 0; i < nums.length; ++i) {
      while (!decreasingDeque.isEmpty() && nums[i] > decreasingDeque.getLast())
        decreasingDeque.removeLast();
      decreasingDeque.addLast(nums[i]);
      if (i >= k - 1) {
        ans.add(decreasingDeque.getFirst());
        if (nums[i - k + 1] == decreasingDeque.getFirst())
          decreasingDeque.removeFirst();
      }
    }

    return ans.stream().mapToInt(i -> i).toArray();
  }
}

File: solutions/cpp/0239.cpp
class Solution {
 public:
  vector<int> maxSlidingWindow(vector<int>& nums, int k) {
    vector<int> ans;
    deque<int> maxQueue;

    for (int i = 0; i < nums.size(); ++i) {
      while (!maxQueue.empty() && maxQueue.back() < nums[i])
        maxQueue.pop_back();
      maxQueue.push_back(nums[i]);
      if (i >= k - 1) {
        ans.push_back(maxQueue.front());
        if (nums[i - k + 1] == maxQueue.front())
          maxQueue.pop_front();
      }
    }

    return ans;
  }
};